### ISSUE TITLE
Issue when using `object` with `Object.entries` and `Object.values`

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [uhyo/better-typescript-lib#40]
[TS Playground](https://www.typescriptlang.org/play/?target=99&ts=5.4.5#code/JYOwLgpgTgZghgYwgAgPICMBWEFgMID2IAzmFAK64FTIDeAUMk8gG5wA25ExAPANLIIAD0ggAJsWQAFKAQAO0MAE8+EJQBpkANQB8ACgIAuZACUc1Mf026AlMa0BtALoBuegF969dhDDIAtkoYmMYEWDh+ALx07m70CESkrBxcktHBEQB0bJzceoHBNm5AA)

As shown above, this overload included in better-typescript-lib cause an issue when applied to an object without a specified key or value type. The overload seems to match too eagerly, returning a type with `never` instead of falling through to the more generic call signatures.

Here's one way to fix it, although I don't know if it's the best fix: [TS Playground](https://www.typescriptlang.org/play/?target=99&ts=5.4.5#code/JYOwLgpgTgZghgYwgAgPICMBWEFgMID2IAzmFAK64FTIDeAUMk8gG5wA25ExAPANLIIAD0ggAJsWQAFKAQAO0MAE8+EJQBpkANQB8ACgIAuZAG0tAXUEiI4ySZAQW0SwH5kDpzWMAlHNTH8mroAlMZaJuYA3PQAvvT07BBgyAC2ShiYxgRYOMkAvHQx0fQIRKSsHFySBRm5AHRsnNx6aRnB0UA)

The same fix seems to work for `Object.entries` as well.

Note: This issue is unrelated to https://github.com/uhyo/better-typescript-lib/pull/31 (but maybe worth mentioning there)

**Comment by ehoogeveen-medweb** [comment]
For posterity and to avoid having to check the playground, the problematic overloads are:
```ts
interface ObjectConstructor {
  values<K extends PropertyKey, V>(o: Record<K, V>): V[];
  entries<K extends PropertyKey, V>(o: Record<K, V>): [string, V][];
}
```
and the proposed fix:
```ts
interface ObjectConstructor {
  values<K extends PropertyKey, V>(o: [V] extends [never] ? never : Record<K, V>): V[];
  entries<K extends PropertyKey, V>(o: [V] extends [never] ? never : Record<K, V>): [string, V][];
}
```

**Comment by ehoogeveen-medweb** [comment]
This might be a better fix, targeting `object` and wider types specifically:
```ts
interface ObjectConstructor {
  values<K extends PropertyKey, V>(o: object extends Record<K, V> ? never : Record<K, V>): V[];
  entries<K extends PropertyKey, V>(o: object extends Record<K, V> ? never : Record<K, V>): [string, V][];
}
```

**Comment by ehoogeveen-medweb** [comment]
Hmm, it seems that `object` and any record with `never`-typed keys mutually extend each other:
```ts
type foo = object extends Record<never, unknown> ? true : false; // true
type bar = Record<never, unknown> extends object ? true : false; // true
```
so to be minimally invasive I guess we want something like
```ts
interface ObjectConstructor {
  values<K extends PropertyKey, V>(o: [K, V] extends [never, never] ? never : Record<K, V>): V[];
  entries<K extends PropertyKey, V>(o: [K, V] extends [never, never] ? never : Record<K, V>): [string, V][];
}
```
so that records with `never`-type keys or `never`-type values still yield the expected result.

Putting it together with https://github.com/uhyo/better-typescript-lib/pull/31:
```ts
type KeyToString<K extends PropertyKey> = K extends string ? K : K extends number ? `${K}` : never;

interface ObjectConstructor {
  keys<K extends PropertyKey, V>(o: [K, V] extends [never, never] ? never : Record<K, V>): KeyToString<K>[];
  values<K extends PropertyKey, V>(o: [K, V] extends [never, never] ? never : Record<K, V>): V[];
  entries<K extends PropertyKey, V>(o: [K, V] extends [never, never] ? never : Record<K, V>): [KeyToString<K>, V][];
}
```
[TS Playground](https://www.typescriptlang.org/play/?target=99&ts=5.4.5#code/C4TwDgpgBA0hIBUD2BlYAnAlgOwOYB4YoIAPYCbAEwGcoAFdJSdUOEAPigF5ZiyKaUahhy4oAfl4AuXqXJVa2AK4BbAEYR0EqAAMAJAG8YAXx1QZ2CADdNAbgBQ9nOXQAzAIYBjaAHk1AKwhPYABhJGxhdCVgpC0DeygoAGt4akI+eUEGJk1WeAAaKAA1dgAKJBkAbRhCooBdDIFaSssbdELWzQbJTq0ZACUg2MpCWvYAShk2ZDQsPEJ2SrqHRKt3ABslCDSiOSb6RmY8kDHyqprihr2FKBbrTQ779G6oXvMoQc9h0eKJmSKlitiNgRNt0tcsodciA2KcKrcLvVGjc7m1Hm0Xm8BkN0CNEX8EfAZiJ5jB2LU6oD7MZHK4lNhgphwlB3NgYfAAHJPcoBbFfXH4bJHdknV5PCZQeKJL4RYDJVLcKB+QLBAB0KRA1B5-nGtigAHp9UISbglgkoDLhFA1pttorlUFgKqbVstUgArqDUbembpeErRRQbQeA61YGsNttZ7DbdIqJ0V1fVB0BBgEp0NhbhrqIUXdtCuHMNtltTafTGczegBVbBJbBIADu2G1fO+vUK9LrjewEqlFv9cuz9oCjvVqSjepjPrq5stcrzwaVI7VC4nXqgnfrTaTc+BQeHKqdhcj7p1k6NqIeG9rW+wlJniRTaYzWdSuY2roLIIj1BLNPsdIMsATKZuQKhgOs7jkFybQth8OIjDorhIEgAC0hhxngphQAAProGEmsYqFqO46A6Am6C9rOA7ypqB6jtma4xqUSEoehBiYbg2F4foHGEcRpE6OMO40Qu9Erh+J4eueYptCJsp7j+4lHt+RZutJ66VKxaEEXMXFmDxumiERJFkRR97mk+6aZpU2bvraOaKWpf5lkBIGyZoCAQOBkHkHBnxtk8hTaexnHcfhfF6SZglUX6ClDiGy5Ooxp7Rt6TzyVaYmJYezqSepZ7rixyE6ZFxkGRFYUCWRwkPv2CnHouoYqUGTEXu2uglaFhEVbxVWmToFmPqm1mvpq9mfk5xYOP+gEVpmygqJomCeDBmj+Qh+AAIy4VAABMu0AMwUbF9VWglS65SlGnMQARHtt27bdW2PXht2HbdtXUQp2WXaOq6pTJ07fQGql2jlo6NW1tz3a9UDPXD723eZSZWS+tlvta+VfkGLkAeWwGVk8HKqMtngbfyIwdTteEHXhh2nbuF3NWOmrQ8DcVZflyl5Q50OlHTUA01Ah1fZzcqNTzUOA5pHWC8Lh1DcmI3o3ZWMOTjP4uUAA)

**Comment by uhyo** [comment]
A fix has been released as `v2.8.0`. Thank you for reporting!
